### PR TITLE
TACO-594 Add publisher implementations

### PR DESCRIPTION
**Pull request by andyjcross**

### CODE REVIEW

File: lib/staged_event/publisher/google_pub_sub.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+require "google/cloud/pubsub"
+
+module StagedEvent
+  module Publisher
+    class GooglePubSub < Base
+      include ShortCircuIt
+
+      def publish(model)
```
Comment (quinn-freshly) [inline lib/staged_event/publisher/google_pub_sub.rb:10]: can you add rdoc for what call signatures are? what is allowed to be a `model`? is the `topic` method provided my a mixin implemented as part of this package?

more to the point, does this expect an object of a specific type, or any object that inherits a specific abstract class or mixin, or any object that has a topic method?

File: spec/lib/staged_event/publisher/google_pub_sub_spec.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    end
+    let(:event_topic) { default_topic }
+    let(:default_topic) { Faker::Lorem.word }
+    let(:alternate_topic) { Faker::Lorem.word }
+    let(:google_pubsub_config) do
+      OpenStruct.new(
+        topic_map: Hash[
+          default_topic, Faker::Lorem.word,
+          alternate_topic, Faker::Lorem.word,
+        ],
```
Comment (quinn-freshly) [inline spec/lib/staged_event/publisher/google_pub_sub_spec.rb:22]:
```
{
          default_topic => Faker::Lorem.word,
          alternate_topic => Faker::Lorem.word,
      }
```

File: lib/staged_event/publisher/google_pub_sub.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+      private
+
+      def google_pubsub
+        Google::Cloud::PubSub.new(
+          project_id: Configuration.config.google_pubsub.project_id,
+          credentials: Configuration.config.google_pubsub.credentials,
+        )
+      end
+      memoize :google_pubsub
```
Comment (quinn-freshly) [inline lib/staged_event/publisher/google_pub_sub.rb:26]: `@attr ||=` seems sufficient here. also, if this is the only place short circuit is used, the dependency could be removed

File: lib/staged_event/publisher/google_pub_sub.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+require "google/cloud/pubsub"
+
+module StagedEvent
+  module Publisher
+    class GooglePubSub < Base
+      include ShortCircuIt
+
+      def publish(model)
+        topic_id = topic_map.fetch(model.topic, topic_map.values.first)
```
Comment (quinn-freshly) [inline lib/staged_event/publisher/google_pub_sub.rb:11]: this seems error-prone. nothing is checking the presence of `topic_map` for having at least one value, and also it silently uses a different topic. I think the behavior should be simplified, and custom errors created for:

* the map is empty
* the map does not contain the desired topic

**Review by quinn-freshly — CHANGES_REQUESTED** [review]

File: lib/staged_event/publisher/google_pub_sub.rb (hunk 3 above)
Comment (andyjcross) [reply lib/staged_event/publisher/google_pub_sub.rb:26]: Ah agreed I was going to make this change 👍

File: lib/staged_event/publisher/google_pub_sub.rb (hunk 1 above)
Comment (andyjcross) [reply lib/staged_event/publisher/google_pub_sub.rb:10]: It's really meant to just be the internal Model class that is backed by the database.
Was going to add rdocs to the `Base` interface when I had a chance 👍

File: lib/staged_event/publisher/google_pub_sub.rb (hunk 4 above)
Comment (andyjcross) [reply lib/staged_event/publisher/google_pub_sub.rb:11]: Ah regarding errors, yeah hadn't quite finished this - at the very least it needs to translate "expected" exceptions from Google into `Publisher::PublishFailedError`, so the publisher process rescues them.

But re: the topic map, in the meantime I was relying on someone running `rails g staged_event::install` and getting the initializer added to their project, which initializes the topic map. Agree it should be more robust 👍

**Comment by andyjcross** [comment]
@quinn-freshly many thanks for looking at these 🙏 it's exceedingly helpful

**Pull request merged**